### PR TITLE
Create saber_BC_buttons.h

### PR DESCRIPTION
**Pull request by NoSloppy**
This is ready for a grueling code review as per suggested on The Crucible.
Now, it does also drag in required edits to saber_base, hybrid_font and voice_data files, so I suppose I also edit those and pull request as well?
Finally, I mistakenly also added this as an upload/ addition/ commit to the OLED pull request. That can be ignored as it's unrelated. Still have GitHub training wheels on.

### CODE REVIEW

**Review by NoSloppy — COMMENTED** [review]
Also changed over SWAP and MONOFORCE to use ready-to-go USER1 and USER2 instead.

File: props/saber_BC_buttons.h [hunk 1]
```diff
@@ -0,0 +1,1092 @@
...
+          hybrid_font.PlayCommon(&SFX_volmin);
+        } else {
+          beeper.Beep(0.5, 1000);
+        }
+        STDOUT.print("Minimum Volume: ");
+      #endif
+      
+    }
+  }
+
```
Comment (profezzorn) [inline props/saber_BC_buttons.h:532]: extra empty line

File: props/saber_BC_buttons.h [hunk 2]
```diff
@@ -0,0 +1,1092 @@
...
+      case EVENTID(BUTTON_POWER, EVENT_PRESSED, MODE_ON):
+      case EVENTID(BUTTON_AUX, EVENT_PRESSED, MODE_ON):
+      case EVENTID(BUTTON_AUX2, EVENT_PRESSED, MODE_ON):
+        if (accel_.x < -0.15) {
+          pointing_down_ = true;
+        } else {
+          pointing_down_ = false;
+        }
+      return true;
+
```
Comment (profezzorn) [inline props/saber_BC_buttons.h:545]: extra empty line

File: props/saber_BC_buttons.h [hunk 3]
```diff
@@ -0,0 +1,1092 @@
...
+      // Due to motion chip startup on boot creating false ignition we delay Swing On at boot for 3000ms
+      if (millis() > 3000) {
+        FastOn();
+  #ifdef GESTURE_AUTO_BATTLE_MODE
+        STDOUT.println("Entering Battle Mode");
+        battle_mode_ = true;
+  #endif
+      }
+      return true;
+  #endif
```
Comment (profezzorn) [inline props/saber_BC_buttons.h:560]: add //  BC_SWING_ON after endif

File: props/saber_BC_buttons.h [hunk 4]
```diff
@@ -0,0 +1,1092 @@
...
+          millis() - saber_off_time_ > 1000) {
+        FastOn();
+  #ifdef GESTURE_AUTO_BATTLE_MODE
+        STDOUT.println("Entering Battle Mode");
+        battle_mode_ = true;
+  #endif
+        last_twist_ = millis();
+      }
+      return true;
+  #endif
```
Comment (profezzorn) [inline props/saber_BC_buttons.h:575]: add // BC_TWIST_ON after endif

File: props/saber_BC_buttons.h [hunk 5]
```diff
@@ -0,0 +1,1092 @@
...
+        Off();
+        last_twist_ = millis();
+        saber_off_time_ = millis();
+        battle_mode_ = false;
+      }
+      return true;
+  #endif
+
+  #ifdef BC_STAB_ON
+        case EVENTID(BUTTON_NONE, EVENT_STAB, MODE_OFF):
```
Comment (profezzorn) [inline props/saber_BC_buttons.h:590]: this case (and the code that goes with it) seem to be indented differently than the previous case and code

File: props/saber_BC_buttons.h [hunk 6]
```diff
@@ -0,0 +1,1092 @@
...
+            } else {
+              hybrid_font.DoEffect(EFFECT_FORCE, 0);
+            }
+            last_push_ = millis();
+          }
+          return true;
+
+  #endif
+
+
```
Comment (profezzorn) [inline props/saber_BC_buttons.h:627]: lots of extra space

File: props/saber_BC_buttons.h [hunk 7]
```diff
@@ -0,0 +1,1092 @@
...
+  #if NUM_BUTTONS == 1
+    case EVENTID(BUTTON_POWER, EVENT_THIRD_SAVED_CLICK_SHORT, MODE_OFF):
+  #else
+    // 2 and 3 button
+    case EVENTID(BUTTON_AUX, EVENT_FIRST_HELD_LONG, MODE_OFF):
+  #endif
+    // accurate battery level reading every time
+      srand((rand() * 917823) ^ analogRead(batteryLevelPin));
+    // needs this delay
+      delay (500);
```
Comment (profezzorn) [inline props/saber_BC_buttons.h:730]: This is not the right way to do this.
Putting delay() in the code causes lots of problems so we shouldn't do that.
Instead we need to put something in loop() that triggers in 500 ms.

File: props/saber_BC_buttons.h [hunk 8]
```diff
@@ -0,0 +1,1092 @@
...
+// Spoken Battery Level
+        
+  #if NUM_BUTTONS == 1
+    case EVENTID(BUTTON_POWER, EVENT_THIRD_SAVED_CLICK_SHORT, MODE_OFF):
+  #else
+    // 2 and 3 button
+    case EVENTID(BUTTON_AUX, EVENT_FIRST_HELD_LONG, MODE_OFF):
+  #endif
+    // accurate battery level reading every time
+      srand((rand() * 917823) ^ analogRead(batteryLevelPin));
```
Comment (profezzorn) [inline props/saber_BC_buttons.h:728]: Why is this here?

File: props/saber_BC_buttons.h [hunk 9]
```diff
@@ -0,0 +1,1092 @@
...
+    // needs this delay
+      delay (500);
+      ::CommandParser::DoParse("play", "battlevel.wav");
```
Comment (profezzorn) [inline props/saber_BC_buttons.h:731]: This isn't the right way to play a wav file.

File: props/saber_BC_buttons.h [hunk 10]
```diff
@@ -0,0 +1,1092 @@
...
+      delay (1900);
+      talkie.SayDigit((int)floorf(battery_monitor.battery()));
+      delay (600);
+      ::CommandParser::DoParse("play", "point.wav");
```
Comment (profezzorn) [inline props/saber_BC_buttons.h:735]: Talkie already has a "point" word.
Seems weird to mix wav file playing and talkie playback. Also if we want to do that we need a better way than just callind delay() to do it.

File: props/saber_BC_buttons.h [hunk 11]
```diff
@@ -0,0 +1,1092 @@
...
+
+// On Demand Battery Level
+
+  case EVENTID(BUTTON_POWER, EVENT_SECOND_SAVED_CLICK_SHORT, MODE_OFF):
+    if (!mode_volume_) {
+      SaberBase::DoEffect(EFFECT_BATTERY_LEVEL, 0);
+    }
+    return true;
+
+
```
Comment (profezzorn) [inline props/saber_BC_buttons.h:752]: extra empty lines

File: props/saber_BC_buttons.h [hunk 12]
```diff
@@ -0,0 +1,1092 @@
...
+  return true;
+
+
+// Blaster Deflection
+
+    // original sa22c mode if ENABLE_AUTO_SWING_BLAST is not defined
+  #ifndef ENABLE_AUTO_SWING_BLAST
+      case EVENTID(BUTTON_POWER, EVENT_FIRST_SAVED_CLICK_SHORT, MODE_ON):
+      case EVENTID(BUTTON_POWER, EVENT_SECOND_SAVED_CLICK_SHORT, MODE_ON):
+              swing_blast_ = false;
```
Comment (profezzorn) [inline props/saber_BC_buttons.h:789]: weird indentation

File: props/saber_BC_buttons.h [hunk 13]
```diff
@@ -0,0 +1,1092 @@
+/*
+saber_BC_buttons.h
+
+Includes 1 and 2 button controls.
+Incorporates an intuitive control scheme so button actions are consistant blade is on or off.
```
[the remark this answers is not in the document]
Comment (NoSloppy) [reply props/saber_BC_buttons.h]: Fixed.

File: props/saber_BC_buttons.h [hunk 14]
```diff
@@ -0,0 +1,1092 @@
+/*
+saber_BC_buttons.h
```
[the remark this answers is not in the document]
Comment (NoSloppy) [reply props/saber_BC_buttons.h:2]: Sure, but I looked at all of the other prop files and don't see anything like that. What would I need that for?

File: props/saber_BC_buttons.h [hunk 15]
```diff
@@ -0,0 +1,1092 @@
...
+#define BC_FORCE_PUSH 
+
+#define BC_FORCE_PUSH_LENGTH
+is used for adjustment to the Push gesture length in millis needed to trigger Force Push
+Recommended range 1 ~ 10, 1 = shortest, easiest to trigger, 10 = longest. Default value is 5.
+
+
+If you want the gesture ignition to ALSO enter battle mode automatically
+on ignition, add this define:
+#define GESTURE_AUTO_BATTLE_MODE
```
[the remark this answers is not in the document]
Comment (NoSloppy) [reply props/saber_BC_buttons.h]: Fixed.

File: props/saber_BC_buttons.h [hunk 16]
```diff
@@ -0,0 +1,1092 @@
...
+
+
+Tightened click timings.
+The timeouts for short and double click detection are shortened 
+from the stock 500ms to 300ms to feel more responsive but still give enough
+timeout to ensure all button actions can be achieved consistently.
+All button timings are included below so they can be easily tweaked to suit
+individual tastes.
+
+
```
[the remark this answers is not in the document]
Comment (NoSloppy) [reply props/saber_BC_buttons.h]: Fixed.

File: props/saber_BC_buttons.h [hunk 17]
```diff
@@ -0,0 +1,1092 @@
...
+EFFECT(volmax);     // for maximum volume reached
+EFFECT(faston);     // for EFFECT_FAST_ON
+EFFECT(blstbgn);    // for Begin Multi-Blast
+EFFECT(blstend);    // for End Multi-Blast
+EFFECT(push);       // for Force Push gesture
+EFFECT(quote);      // for playing quotes
+EFFECT(monosfx);    // for Monophonically played sounds (iceblade, seismic charge etc...)
+EFFECT(swap);       // for standalone triggering EffectSequence<>
+
+
```
[the remark this answers is not in the document]
Comment (NoSloppy) [reply props/saber_BC_buttons.h]: Fixed. All of the following same "too many empty lines" also fixed instead of replying to all of them.

File: props/saber_BC_buttons.h [hunk 18]
```diff
@@ -0,0 +1,1092 @@
...
+          fabs(fusor.gyro().x) < 10) {
+        if (millis() - push_begin_millis_ > BC_FORCE_PUSH_LENGTH) {
+          Event(BUTTON_NONE, EVENT_PUSH);
+          push_begin_millis_ = millis();
+        }
+      } else {
+        push_begin_millis_ = millis();
+      }
+      
+    // EVENT_THRUST
```
[the remark this answers is not in the document]
Comment (NoSloppy) [reply props/saber_BC_buttons.h]: I changed it to one instance and just moved it out of the SaberBase::IsOn() condition. Better?

File: props/saber_BC_buttons.h (hunk 5 above)
Comment (NoSloppy) [reply props/saber_BC_buttons.h]: Fixed, all identical now.

File: props/saber_BC_buttons.h (hunk 7 above)
Comment (NoSloppy) [reply props/saber_BC_buttons.h]: Removed, see below.

File: props/saber_BC_buttons.h (hunk 8 above)
Comment (NoSloppy) [reply props/saber_BC_buttons.h]: Old remnant from back in like 2018 I think. The explanation I got was "waits for the battery entropy to even out. So its like poking the tiger before you test for battery level so its accurate. Without that, you can get a different result if you test it twice in a row"
It never hurt anything so I've left in in through all the updates.
Removed it and the first delay it apparently needed.

File: props/saber_BC_buttons.h (hunk 9 above)
Comment (NoSloppy) [reply props/saber_BC_buttons.h]: I would love to do it the right way. I changed them to be like vmbegin and all of those. Better?

File: props/saber_BC_buttons.h (hunk 10 above)
Comment (NoSloppy) [reply props/saber_BC_buttons.h]: Sure, would wait() be better than delay? Not sure how else to make the timing work.

I know it seems weird, but I was just using as much wav as possible for sonic clarity. I tried both ways, and if you actually listen to it it seems more congruent using the point.wav. Either way though, with OS6 pending and the ability to have wavs used for all the digits instead of talkie, would it be moot to even pursue this whole section now?

That being said, I made the wavs conditional on existing.

File: props/saber_BC_buttons.h (hunk 12 above)
Comment (NoSloppy) [reply props/saber_BC_buttons.h]: Fixed all incorrect indentation throughout the file. Let me know if I missed anything.

File: props/saber_BC_buttons.h (hunk 10 above)
Comment (profezzorn) [reply props/saber_BC_buttons.h]: I'm not sure what wait() is, but any function that just sits around and does nothing is not going to work. ProffieOS has better things to do than sitting around and do nothing.

To "wait" for something, you have to build something that polls for it. Usually this means putting something in loop() that checks if it's been long enough to do the next thing.

ProffieOS also have these lightweight threads that can make it look like you can wait for things, but it's all fakery and they end up actually do the same kind of polling that I described above.

File: props/saber_BC_buttons.h (hunk 14 above)
[the remark this answers is not in the document]
Comment (NoSloppy) [reply props/saber_BC_buttons.h:2]: Any pointers here?

[the remark this answers is not in the document]
Comment (NoSloppy) [reply props/saber_BC_buttons.h:2]: Fredrik, when you get a chance, this one has me stumped.

[the remark this answers is not in the document]
Comment (profezzorn) [reply props/saber_BC_buttons.h:2]: It can be pretty simple, something like:

Copyright 2021 by:
... list of authors here
Distributed under the terms of the GNU General Public License v3.

File: props/saber_BC_buttons.h [hunk 19]
```diff
@@ -0,0 +1,1046 @@
...
+// Spoken Battery Level
+  #if NUM_BUTTONS == 1
+    case EVENTID(BUTTON_POWER, EVENT_THIRD_SAVED_CLICK_SHORT, MODE_OFF):
+  #else
+    // 2 and 3 button
+    case EVENTID(BUTTON_AUX, EVENT_FIRST_HELD_LONG, MODE_OFF):
+  #endif 
+      if (SFX_battlevel) {        
+        hybrid_font.PlayCommon(&SFX_battlevel);
+        talkie.Say(spPAUSE1);
```
Comment (profezzorn) [inline props/saber_BC_buttons.h:704]: This is bad.
I recommend writing this with a state machine instead.

File: props/saber_BC_buttons.h (hunk 14 above)
[the remark this answers is not in the document]
Comment (NoSloppy) [reply props/saber_BC_buttons.h:2]: Updated.

File: props/saber_BC_buttons.h (hunk 19 above)
Comment (NoSloppy) [reply props/saber_BC_buttons.h]: Bailed on the spoken stuff in light of pending OS6 edit mode.
Straight talkie now.
